packages/@aws-cdk/aws-iam/lib/private/merge-statements.ts

```diff
@@ -0,0 +1,223 @@
+        break;
+      }
+    }
+
+    if (!didMerge) {
+      i++;
+    }
+  }
+
+  return compStatements.map(renderComparable);
+}
+
+/**
+ * Given two statements, return their merging (if possible)
+ *
+ * We can merge two statements if:
+ *
+ * - Their effects are the same
+ * - They don't have Sids (not really a hard requirement, but just a simplification and an escape hatch)
+ * - Their Conditions are the same
+ * - Their NotAction, NotResource and NotPrincipal sets are the same (empty sets is fine).
+ * - From their Action, Resource and Principal sets, 2 are subsets of each other
+ *   (empty sets are fine).
+ */
+function tryMerge(a: ComparableStatement, b: ComparableStatement): ComparableStatement | undefined {
+  // Effects must be the same
+  if (a.effect !== b.effect) { return; }
+  // We don't merge Sids (for now)
+  if (a.sid || b.sid) { return; }
+
+  if (a.conditionString !== b.conditionString) { return; }
+  if (!setEqual(a.notAction, b.notAction) || !setEqual(a.notResource, b.notResource) || !setEqual(a.notPrincipal, b.notPrincipal)) { return; }
+
+  // We can merge these statements if 2 out of the 3 sets of Action, Resource, Principal
+  // are the same.
+  const setsEqual = (setEqual(a.action, b.action) ? 1 : 0) +
+    (setEqual(a.resource, b.resource) ? 1 : 0) +
+    (setEqual(a.principal, b.principal) ? 1 : 0);
+
+  if (setsEqual < 2) { return; }
+
+  return {
+    effect: a.effect,
+    conditionString: a.conditionString,
+    conditionValue: b.conditionValue,
+    notAction: a.notAction,
+    notPrincipal: a.notPrincipal,
+    notResource: a.notResource,
+
+    action: setMerge(a.action, b.action),
+    resource: setMerge(a.resource, b.resource),
+    principal: setMerge(a.principal, b.principal),
+  };
+
+  function setMerge(x: IamValueSet, y: IamValueSet): IamValueSet {
+    return { ...x, ...y };
+  }
+}
+
+/**
+ * Calculate and return cached string set representation of the statement elements
+ *
+ * This is to be able to do comparisons on these sets quickly.
+ */
+function makeComparable(s: StatementSchema): ComparableStatement {
+  return {
+    effect: s.Effect,
+    sid: s.Sid,
+    action: iamSet(s.Action),
+    notAction: iamSet(s.NotAction),
+    resource: iamSet(s.Resource),
+    notResource: iamSet(s.NotResource),
+    principal: principalIamSet(s.Principal),
+    notPrincipal: principalIamSet(s.NotPrincipal),
+    conditionString: JSON.stringify(s.Condition),
+    conditionValue: s.Condition,
+  };
+
+  function forceArray<A>(x: A | Array<A>): Array<A> {
+    return Array.isArray(x) ? x : [x];
+  }
+
+  function iamSet(x: IamValue | undefined): IamValueSet {
+    if (x == undefined) { return {}; }
+    return mkdict(forceArray(x).map(e => [JSON.stringify(e), e]));
+  }
+
+  function principalIamSet(x: IamValue | Record<string, IamValue> | undefined): IamValueSet {
+    if (x === undefined) { return {}; }
+
+    if (Array.isArray(x) || typeof x === 'string') {
+      x = { AWS: x };
+    }
```

Comment:
Why are we performing normalization here / can we normalize the statements in this way before they are passed to `mergeStatements`? I feel like normalization and statement merging should be logically separate steps, to the extent that it's possible anyways.

Reply:
This is not where we merge. This is where we "lift" the IAM statement into a representation that is amenable to merging. Is this not enough separation?